api/api/routers/mcp/_mcp_observability_middleware.py

```diff
@@ -0,0 +1,341 @@
+                result=tool_result,
+                started_at=_tool_call_start_times.get(request_id, tool_call_start_time),
+                completed_at=completed_at,
+                user_agent=user_agent,
+            )
+
+            # Update session state with completed tool call (async call in sync context)
+            async def update_session():
+                await session_state.register_tool_call(tool_call_data)
+
+            # Clean up start time tracking
+            _tool_call_start_times.pop(request_id, None)
+
+            # Run observer agent in background
+            if response_body:
+                observer_data = ObserverAgentData(
+                    tool_name=tool_name,
+                    previous_tool_calls=session_state.tool_calls,
+                    tool_arguments=tool_arguments,
+                    tool_result=tool_result,
+                    duration_seconds=duration,
+                    user_agent=user_agent,
+                    mcp_session_id=session_state.session_id,
+                    request_id=str(request_id),
+                    organization_name=tenant_info.slug,
+                )
+
+                # Create background task for observer agent
+                async def run_observer_background():
+                    # Update session first, then run observer
+                    await update_session()
+                    await self._run_observer_agent_background(observer_data)
+
+                add_background_task(run_observer_background())
+            else:
+                # Just update session if no response body
+                add_background_task(update_session())
+                logger.warning(
+                    "No response body received, skipping observer agent",
+                    extra={
+                        "session_id": session_state.session_id,
+                        "request_id": request_id,
+                    },
+                )
+
+        return on_streaming_complete
+
+    async def dispatch(self, request: Request, call_next: Callable[[Request], Any]) -> Response:
+        """Main dispatch method with comprehensive error handling"""
+        # Setup session and validate request
+        setup_result = await self._setup_session_and_validate_request(request)
+        if not setup_result:
+            return await call_next(request)
+
+        tenant_info, session_state, request_data = setup_result
+
+        # Validate MCP tool call and extract info
+        tool_call_info = await self._validate_mcp_tool_call(request_data, session_state)
+        if not tool_call_info:
+            return await call_next(request)
+
+        tool_name, tool_arguments, request_id = tool_call_info
+
+        # Track timing
+        start_time = time.time()
+        user_agent = request.headers.get("user-agent", "unknown")
+        tool_call_start_time = datetime.now(timezone.utc)
+
+        try:
+            _tool_call_start_times[request_id] = tool_call_start_time
```

Comment:
Why would that raise ?

Reply:
Also why store it in a global variable ? Can't you just pass it along ?

Reply:
removed https://github.com/WorkflowAI/WorkflowAI/pull/571/commits/75049fc5a75487e916f479b442d064fa5e71d11f